snapcraft/plugins/_ros/rospack.py

```diff
@@ -0,0 +1,146 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2019 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import logging
+import subprocess
+import tempfile
+
+from snapcraft.internal import common, repo
+from snapcraft import formatting_utils
+
+logger = logging.getLogger(__name__)
+
+
+class Rospack:
+    def __init__(
+        self,
+        *,
+        ros_distro,
+        ros_package_path,
+        rospack_path,
+        ubuntu_sources,
+        ubuntu_keyrings,
+        project
+    ):
+        self._ros_distro = ros_distro
+        self._ros_package_path = ros_package_path
+        self._rospack_path = rospack_path
+        self._ubuntu_sources = ubuntu_sources
+        self._ubuntu_keyrings = ubuntu_keyrings
+        self._project = project
+
+        self._rospack_install_path = os.path.join(self._rospack_path, "install")
+        self._rospack_cache_path = os.path.join(self._rospack_path, "cache")
+
+    def setup(self):
+        # Make sure we can run multiple times without error
+        os.makedirs(self._rospack_install_path, exist_ok=True)
+
+        # rospack isn't necessarily a dependency of the project, so we'll unpack
+        # it off to the side and use it from there.
+        logger.info("Preparing to fetch rospack...")
+        ubuntu = repo.Ubuntu(
+            self._rospack_path,
+            sources=self._ubuntu_sources,
+            keyrings=self._ubuntu_keyrings,
+            project_options=self._project,
+        )
+
+        logger.info("Fetching rospack...")
+        ubuntu.get(
+            [
+                "ros-{}-rospack".format(self._ros_distro),
+                "ros-{}-catkin".format(self._ros_distro),
+            ]
+        )
+
+        logger.info("Installing rospack...")
+        ubuntu.unpack(self._rospack_install_path)
+
+    def list_names(self):
+        """Obtain list of packages present in the workspace."""
+        output = self._run(["list-names"]).strip()
+        if output:
+            return set(output.split("\n"))
+        else:
+            return set()
+
+    def _run(self, arguments):
+        with tempfile.NamedTemporaryFile(mode="w+") as f:
+            lines = [
+                'export PYTHONPATH="{}"'.format(
+                    os.path.join(
+                        self._rospack_install_path,
+                        "usr",
+                        "lib",
+                        "python2.7",
```

Comment:
Is this always going to be python2.7?

Reply:
Yeah, python2 is EOL at 2.7 (even if it bumped, it won't be SRU'd back to old Ubuntu releases).